textx/model.py

```diff
@@ -272,23 +271,108 @@ def get_model_from_str(self, model_str, file_name=None, debug=None,
                     self.dprint("*** PARSING MODEL ***")
 
                 self.parse(model_str, file_name=file_name)
+
+                # Used to keep track of user class instances
+                self._user_class_inst = []
+
+                self._replace_user_attr_methods()
+
                 # Transform parse tree to model. Skip root node which
                 # represents the whole file ending in EOF.
                 model = parse_tree_to_objgraph(
                     self, self.parse_tree[0], file_name=file_name,
                     pre_ref_resolution_callback=pre_ref_resolution_callback,
                     is_main_model=is_main_model, encoding=encoding)
+
+            except BaseException as e:
+                # Restore of user classes replaced attr methods
+                self._restore_user_attr_methods()
+                raise e
+
             finally:
+
                 if debug is not None:
                     self.debug = old_debug_state
 
-            try:
-                model._tx_metamodel = self.metamodel
-            except AttributeError:
-                # model is some primitive python type (e.g. str)
-                pass
             return model
 
+        # Custom attr dunder methods used for user classes during loading
+        def _getattribute(obj, name):
+            if name == '__dict__':
+                try:
+                    return type(obj)._tx_obj_attrs[id(obj)]
+                except KeyError:
+                    pass
+            return super(type(obj), obj).__getattribute__(name)
+
+        def _getattr(obj, name):
+            try:
+                return obj._tx_obj_attrs[id(obj)][name]
+            except KeyError:
+                try:
+                    return obj._tx_real_getattr(name)
+                except (AttributeError, TypeError):
+                    return super(type(obj), obj).__getattr__(name)
+
+        def _setattr(obj, name, value):
+            try:
+                obj._tx_obj_attrs[id(obj)][name] = value
+            except KeyError:
+                try:
+                    return obj._tx_real_setattr(name, value)
+                except (AttributeError, TypeError):
+                    return super(type(obj), obj).__setattr__(name, value)
+
+        def _delattr(obj, name):
+            try:
+                obj._tx_obj_attrs[id(obj)].pop(name)
+            except KeyError:
+                try:
+                    return obj._tx_real_delattr(name)
+                except (AttributeError, TypeError):
+                    return super(type(obj), obj).__delattr__(name)
+
+        def _replace_user_attr_methods(self):
+            """
+            Replace get/set/del(attr) methods on user classes
+            to support postponing of user obj initialization.
+            """
+            for user_class in self.metamodel.user_classes.values():
+                if not hasattr(user_class, '_tx_instrumented'):
```

Comment:
If a user class is a sub class of another user class, which has already been instrumented, then this returns true, even though the class itself still uses its original `attr` methods. I fixed this in https://github.com/textX/textX/pull/258 by checking for `if '_tx_instrumented' not in user_class.__dict__:`